cmd/kubeadm/app/master/tokens.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/api"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/util/uuid"
+)
+
+func generateTokenIfNeeded(s *kubeadmapi.KubeadmConfig) error {
+	ok, err := kubeadmutil.UseGivenTokenIfValid(s)
+	if !ok {
+		if err != nil {
+			return err
+		}
+		err = kubeadmutil.GenerateToken(s)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("<master/tokens> generated token: %q\n", s.Secrets.GivenToken)
+	} else {
+		fmt.Println("<master/tokens> accepted provided token")
+	}
+
+	return nil
+}
+
+func CreateTokenAuthFile(s *kubeadmapi.KubeadmConfig) error {
```

Comment:
Perhaps a better overall flow would be

```go
func CreateTokenAuthFile(s *kubeadmapi.KubeadmConfig) error {
	if s.Secrets.GivenToken == "" {
		s.Secrets = kubeadmutil.GenerateToken()
	} else {
		s.Secrets = kubeadmutil.ParseToken(s.Secrets.GivenToken)
	}
	// the rest of the file handling
}
```

(error handling omitted)

Then you wouldn't need `generateTokenIfNeeded`.